pkg/liqo-controller-manager/authentication/virtualnodecreator-controller/virtualnodecreator_controller.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/liqotech/liqo/pkg/liqo-controller-manager/authentication"
+	"github.com/liqotech/liqo/pkg/liqo-controller-manager/offloading/forge"
+	"github.com/liqotech/liqo/pkg/utils/getters"
+)
+
+// VirtualNodeCreatorReconciler create virtualnodes from resourceslice resources.
+type VirtualNodeCreatorReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	EventRecorder record.EventRecorder
+}
+
+// NewVirtualNodeCreatorReconciler returns a new VirtualNodeCreatorReconciler.
+func NewVirtualNodeCreatorReconciler(cl client.Client, s *runtime.Scheme, recorder record.EventRecorder) *VirtualNodeCreatorReconciler {
+	return &VirtualNodeCreatorReconciler{
+		Client: cl,
+		Scheme: s,
+
+		EventRecorder: recorder,
+	}
+}
+
+// cluster-role
+// +kubebuilder:rbac:groups=virtualkubelet.liqo.io,resources=virtualnodes,verbs=get;list;watch;delete;create;update;patch
+// +kubebuilder:rbac:groups=authentication.liqo.io,resources=resourceslices,verbs=get;list;watch
+// +kubebuilder:rbac:groups=authentication.liqo.io,resources=identities,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;delete;create;update;patch
+
+// Reconcile resourceslices and create their associated virtualnodes.
+func (r *VirtualNodeCreatorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var resourceSlice authv1alpha1.ResourceSlice
+	if err := r.Get(ctx, req.NamespacedName, &resourceSlice); err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(4).Infof("resourceSlice %q not found", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("unable to get ResourceSlice %q: %v", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	}
+
+	if resourceSlice.Annotations == nil ||
+		resourceSlice.Annotations[consts.CreateVirtualNodeAnnotation] == "" ||
+		strings.EqualFold(resourceSlice.Annotations[consts.CreateVirtualNodeAnnotation], "false") {
+		klog.V(6).Infof("VirtualNode creation disabled for resourceslice %q", req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	if resourceSlice.Labels == nil || resourceSlice.Labels[consts.RemoteClusterID] == "" {
+		err := fmt.Errorf("resourceslice %q does not contain the remote cluster ID label", req.NamespacedName)
+		klog.Error(err)
+		return ctrl.Result{}, err
+	}
+
+	remoteClusterID := resourceSlice.Labels[consts.RemoteClusterID]
+
+	// Get the associated Identity for the remote cluster.
+	identity, err := getters.GetIdentityFromResourceSlice(ctx, r.Client, remoteClusterID, resourceSlice.Name)
+	if err != nil {
+		klog.Errorf("Unable to get the Identity associated to resourceslice %q: %s", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	}
+
+	// Get associated secret
+	if identity.Status.KubeconfigSecretRef == nil || identity.Status.KubeconfigSecretRef.Name == "" {
+		err := fmt.Errorf("identity %q does not contain the kubeconfig secret reference", identity.Name)
+		klog.Error(err)
+		return ctrl.Result{}, err
```

Comment:
I would avoid returning errors here since it will error forever until other controllers fill the status.

It is more correct to return a nil error here without re-enqueue and set an own in the setupwithmanager function to trigger reconciliations on identity updates